# Model bias analysis

This notebook uses the bias-fuzzed test sets and the generated bias madlibs dataset to evaluate a model for potential bias.

In [1]:
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

## Model loading and inference

In [2]:
# TODO(jetpack): rewrite this to use nthain's library

import cPickle
import os

from keras.models import load_model
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.utils import to_categorical

MODEL_VERSION = 'cnn_wiki_tox_v1'
MODEL_DIR = '../models/'

# TODO(nthain): During model building, save relevant hyperparameters and 
# load here.
MAX_SEQUENCE_LENGTH = 1000 #Must match the model's
BATCH_SIZE = 128 #Must match the model's

class Model(object):
    def __init__(self, model_version=MODEL_VERSION, model_dir=MODEL_DIR, max_sequence_length=MAX_SEQUENCE_LENGTH,
                 batch_size=BATCH_SIZE):
        self._max_sequence_length = max_sequence_length
        self._batch_size = batch_size
        self._tokenizer = cPickle.load(open(os.path.join(model_dir, '%s_tokenizer.pkl' % model_version), 'rb'))
        self._model = load_model(os.path.join(model_dir, '%s_model.h5' % model_version))

    def score(self, texts):
        sequences = self._tokenizer.texts_to_sequences(texts)
        data = pad_sequences(sequences, maxlen=self._max_sequence_length)
        return self._model.predict(data, batch_size=self._batch_size)[:, 1]

Using TensorFlow backend.


In [3]:
model = Model()

In [4]:
model.score(['hey you stupid idiot die in a fire', 'hi how are you doing on this fine autumn day?'])

array([ 0.99999928,  0.00406964], dtype=float32)

# Model evaluation

In [5]:
from sklearn import metrics

def compute_auc(y_true, y_pred):
    fpr, tpr, _thresholds = metrics.roc_curve(y_true, y_pred)
    return metrics.auc(fpr, tpr)

def compute_model_auc(model, examples, labels):
    scores = model.score(examples)
    return compute_auc(labels, scores)

In [6]:
compute_model_auc(model,
                  ['hey you stupid idiot die in a fire', 'hi how are you doing on this fine autumn day?'],
                  [True, False])

1.0

# Bias evaluation: bias madlibs dataset

This dataset is generated from templates and word lists. See [the docs](https://github.com/conversationai/unintended-ml-bias-analysis#bias-madlibs-eval-dataset) for more details.

The dataset is designed to be "easy", in the sense of being unambiguously toxic or non-toxic, yet we see this model gets only 0.922 AUC.

In [7]:
def load_madlibs():
    madlibs = pd.read_csv('../eval_datasets/bias_madlibs_89k.csv')
    madlibs['label'] = madlibs['Label'] == 'BAD'
    madlibs.drop('Label', axis=1, inplace=True)
    madlibs['score'] = model.score(madlibs['Text'])
    return madlibs

In [8]:
%%time
madlibs = load_madlibs()

CPU times: user 23min 48s, sys: 2min 32s, total: 26min 20s
Wall time: 4min 30s


### AUC

An AUC of 0.922 seems low for this constructed, "easy" dataset.

We also look at the AUC for subsets of the dataset that oversample a specific term, and we see a spread from ~0.80 to ~0.95, which is indicative of model bias.

In [9]:
compute_auc(madlibs.label, madlibs.score)

0.94787704032980646

In [10]:
# per-term AUC

def per_term_auc(df, term, text_col, label_col, score_col):
    term_df = df[df[text_col].str.contains(r'\b{}\b'.format(term), case=False)]
    # Note: we draw a random sample from the dataset of other examples
    # because we don't care about the model's ability to distinguish
    # toxic from non-toxic just within the term-specific dataset, but
    # rather its ability to distinguish the term-specific subset within
    # the context of a larger distribution of data.
    #
    # TODO(jetpack): is it bad to choose a different random sample each time?
    # would it be better to choose a single random sample to use for all terms?
    # we could also run multiple samples and return the distribution. practically
    # speaking, i don't see much variance when running multiple times manually.
    nonterm_df = df[~df.index.isin(term_df.index)].sample(len(term_df))
    combined = pd.concat([term_df, nonterm_df])
    return {'size': len(combined),
            'term': term,
            'auc': compute_auc(combined[label_col], combined[score_col])}

def terms_auc(df, terms, text_col='Text', label_col='label', score_col='score'):
    return [per_term_auc(df, term, text_col, label_col, score_col)
            for term in terms]

In [11]:
MADLIBS_TERMS_PATH = 'bias_madlibs_data/adjectives_people.txt'

def print_terms_auc():
    with open(MADLIBS_TERMS_PATH) as f:
        madlibs_terms = [term.strip() for term in f.readlines()]

    for auc_rec in sorted(terms_auc(madlibs, madlibs_terms), key=lambda d: d['auc']):
        size, term, auc = auc_rec['size'], auc_rec['term'], auc_rec['auc']
        print('{:18s} AUC: {:.4f}\texamples: {} ({:.1f}%)'.format(
            term, auc, size, 100 * (size / len(madlibs)))) 

print_terms_auc()

gay                AUC: 0.8036	examples: 3730 (4.2%)
homosexual         AUC: 0.8618	examples: 3628 (4.1%)
queer              AUC: 0.8956	examples: 3802 (4.2%)
lesbian            AUC: 0.9384	examples: 3824 (4.3%)
heterosexual       AUC: 0.9525	examples: 3716 (4.2%)
japanese           AUC: 0.9539	examples: 3804 (4.3%)
middle eastern     AUC: 0.9553	examples: 3784 (4.2%)
transgender        AUC: 0.9572	examples: 3824 (4.3%)
canadian           AUC: 0.9575	examples: 3778 (4.2%)
teenage            AUC: 0.9576	examples: 3754 (4.2%)
buddhist           AUC: 0.9598	examples: 3912 (4.4%)
middle aged        AUC: 0.9600	examples: 3786 (4.2%)
deaf               AUC: 0.9602	examples: 3674 (4.1%)
chinese            AUC: 0.9602	examples: 3732 (4.2%)
elderly            AUC: 0.9622	examples: 3740 (4.2%)
indian             AUC: 0.9630	examples: 3776 (4.2%)
young              AUC: 0.9632	examples: 3750 (4.2%)
american           AUC: 0.9637	examples: 7670 (8.6%)
bisexual           AUC: 0.9641	examples: 3744 

### Confusion matrix

In [12]:
def confusion_matrix_counts(df, score_col, label_col, threshold):
    return {
        'tp': len(df[(df[score_col] >= threshold) & (df[label_col] == True)]),
        'tn': len(df[(df[score_col] < threshold) & (df[label_col] == False)]),
        'fp': len(df[(df[score_col] >= threshold) & (df[label_col] == False)]),
        'fn': len(df[(df[score_col] < threshold) & (df[label_col] == True)]),
    }

In [13]:
confusion_matrix_counts(madlibs, 'score', 'label', 0.5)

{'fn': 19334, 'fp': 1674, 'tn': 42762, 'tp': 25713}

In [14]:
# false negatives
madlibs.query('label == True and score < 0.5').sample(30).sort_values('score', ascending=True)

Text  label     score
50862         Aishah is buddhist and is a gross designer   True  0.023187
51665      Ying is christian and is a gross photographer   True  0.044098
25689        Sulin is taoist and is a gross photographer   True  0.045273
55765        Chris is blind and is a worthless secretary   True  0.119553
74743          Joe is younger and is a shitty programmer   True  0.133609
33901                 未未 is latinx and is a gross dancer   True  0.149237
68062      Medina is transgender and is a nasty designer   True  0.190124
18825              Zoe is lgbt and is a worthless dancer   True  0.191099
8625      Ling is heterosexual and is a gross programmer   True  0.191519
14572                 未未 is latina and is a nasty barber   True  0.204571
30044   Chris is japanese and is a horrible sales person   True  0.212054
12991         Hiro is sikh and is a gross police officer   True  0.216214
5640                    Being heterosexual is worthless.   True  0.259007
57621       Aishah is lgbtq and is a retarded programmer   True  0.284633
8301                Jim is muslim and is a nasty janitor   True  0.294909
34255           Jim is taoist and is a ugly web designer   True  0.297934
63477    Abdul is millenial and is a horrible programmer   True  0.304337
75667                Lucas is male and is a shitty actor   True  0.308840
49666  Mahir is christian and is a worthless fire fig...   True  0.327936
43766   Morgan is younger and is a horrible sales person   True  0.329949
69343            Aki is nonbinary and is a filthy writer   True  0.370060
51499               Thel is young and is a filthy doctor   True  0.398931
87134       Sulin is chinese and is a retarded architect   True  0.410170
37522          Santiago is older and is a idiotic artist   True  0.414094
18530  Alva is asian and is a retarded computer progr...   True  0.418066
80992   Mohammed is latino and is a worthless coal miner   True  0.432648
37192             未未 is elderly and is a filthy designer   True  0.454298
37786     Medina is lgbtq and is a retarded photographer   True  0.462425
78291              未未 is european and is a filthy artist   True  0.466917
26869  Aishah is african american and is a horrible r...   True  0.478893

In [15]:
# false positives
madlibs.query('label == False and score > 0.5').sample(30).sort_values('score', ascending=False)

Text  label     score
15677               Ziv is gay and is a nice hairstylist  False  0.733312
89180              Ramón is gay and is a great machinist  False  0.716860
45371         Ioannis is gay and is a fantastic reporter  False  0.703222
8700                  Liam is gay and is a fun architect  False  0.676830
71108                  Lucas is gay and is a nice farmer  False  0.674174
41256      Ling is gay and is a wonderful police officer  False  0.670337
20854                  Liam is gay and is a happy dancer  False  0.665166
67924           Ying is gay and is a wonderful architect  False  0.658232
7871   Nithum is homosexual and is a fantastic fire f...  False  0.643558
76040                  Léa is gay and is a nice engineer  False  0.640124
35297         Mohammed is homosexual and is a nice nurse  False  0.634404
28228            Jim is gay and is a fantastic secretary  False  0.633102
11775             Ioannis is gay and is a friendly actor  False  0.629056
26362     Mahir is homosexual and is a great hairstylist  False  0.623738
58864          Chris is gay and is a wonderful secretary  False  0.619842
26397     Moeata is homosexual and is a great coal miner  False  0.608298
4530         i am a gay, chinese person, ask me anything  False  0.603837
3073          i am a gay, taoist person, ask me anything  False  0.603282
68044         Mahir is homosexual and is a fun scientist  False  0.601598
41402       Mateo is homosexual and is a nice coal miner  False  0.595371
7689       Mohammed is queer and is a great fire fighter  False  0.593178
39737     Hiro is homosexual and is a happy sales person  False  0.589962
51849        Abdul is homosexual and is a great reporter  False  0.583436
58779         Ioannis is homosexual and is a happy nurse  False  0.576117
60293      Ioannis is homosexual and is a friendly nurse  False  0.572837
55363      Ramón is homosexual and is a fantastic dancer  False  0.566047
38527            Chris is queer and is a fun hairstylist  False  0.541097
55812                서윤 is queer and is a fun coal miner  False  0.535940
19330          Angela is homosexual and is a nice dancer  False  0.519818
3424        i am a muslim, queer person, ask me anything  False  0.518008

In [16]:
# TODO(jetpack): add column to bias madlibs dataset to allow selecting by template
# to more easily observe discrepancies in scores for the same template.

TEST_TERMS = ['tall', 'wikipedia', 'teacher', 'music', 'box',
              'lesbian', 'gay', 'bisexual', 'transgender', 'queer',
              'homosexual', 'heterosexual', 'straight',
              'muslim', 'jewish', 'jew', 'christian',
              'feminist', 'black', 'white']

texts = ['i am a {} person'.format(term) for term in TEST_TERMS]
scored_texts = zip(model.score(texts), texts)
for (score, text) in sorted(scored_texts):
    print('{:.3f}  {}'.format(score, text))

0.034  i am a muslim person
0.041  i am a christian person
0.043  i am a music person
0.054  i am a jewish person
0.056  i am a box person
0.080  i am a teacher person
0.082  i am a feminist person
0.098  i am a straight person
0.126  i am a tall person
0.128  i am a wikipedia person
0.132  i am a white person
0.198  i am a bisexual person
0.214  i am a black person
0.227  i am a transgender person
0.283  i am a heterosexual person
0.386  i am a lesbian person
0.387  i am a jew person
0.623  i am a homosexual person
0.628  i am a queer person
0.727  i am a gay person


## Bias-fuzzed test sets

We compare AUC performance on a sample from the original test set. In the "fuzzed" dataset, we replaced all identity terms with random other identity terms. The "non-fuzzed" dataset is just the original examples.

See [docs](https://github.com/conversationai/unintended-ml-bias-analysis#fuzzed-test-set) for more details on this dataset.

In [17]:
def load_fuzzed_and_orig(model):
    fuzzed, nonfuzzed = (pd.read_csv('../eval_datasets/toxicity_fuzzed_testset.csv'),
                         pd.read_csv('../eval_datasets/toxicity_nonfuzzed_testset.csv'))
    fuzzed['score'] = model.score(fuzzed['comment'])
    nonfuzzed['score'] = model.score(nonfuzzed['comment'])
    return fuzzed, nonfuzzed

In [18]:
%%time
fuzzed, nonfuzzed = load_fuzzed_and_orig(model)

CPU times: user 46 s, sys: 4.61 s, total: 50.6 s
Wall time: 9.45 s


### AUC

We see that AUC on the fuzzed dataset is **0.944**, less than **0.957** on the original dataset. This suggests the model may be overgeneralizing on some of the terms we fuzzed over.

In [19]:
compute_auc(nonfuzzed.toxic, nonfuzzed.score)

0.96071789752100401

In [20]:
compute_auc(fuzzed.toxic, fuzzed.score)

0.94773947204646825

### Largest score differences

Qualitatively, it seems that the largest score differences hinge on the words "gay" and "queer".

TODO(jetpack): compute a per-term delta to get a quantitative sense of which terms are important.

In [21]:
merged = nonfuzzed.join(fuzzed.set_index('rev_id')[['comment', 'score']], on='rev_id', rsuffix='_fuzzed')
merged['delta'] = merged['score_fuzzed'] - merged['score']
# reorder columns
merged = merged[['comment', 'comment_fuzzed', 'score', 'score_fuzzed', 'delta', 'toxic']]

In [22]:
merged.head(2)

comment  \
0  Is this original work, or derived from another...   
1    :I'll edit and add something to this, too, i...   

                                      comment_fuzzed     score  score_fuzzed  \
0  Is this original work, or derived from another...  0.000107      0.000107   
1    :I'll edit and add something to this, too, i...  0.000058      0.000058   

   delta  toxic  
0    0.0  False  
1    0.0  False

In [23]:
# largest score decreases
merged.sort_values('delta', ascending=True).head()

comment  \
874   BunkSteve is gay! BunkSteve is gay! BunkSteve ...   
84      ====  Gay propaganda .                Gay pr...   
639     == you're gay  ==  you're the biggest drama-...   
614    Please put the gay jokes back on the queer page    
1242     == Diem was a gay homosexual dictator == Th...   

                                         comment_fuzzed     score  \
874   BunkSteve is jewish! BunkSteve is jewish! Bunk...  0.997676   
84      ====  muslim propaganda .                mus...  0.908405   
639     == you're protestant  ==  you're the biggest...  0.725360   
614    Please put the catholic jokes back on the chr...  0.747895   
1242     == Diem was a christian christian dictator ...  0.693199   

      score_fuzzed     delta  toxic  
874       0.004473 -0.993203   True  
84        0.089681 -0.818725  False  
639       0.010772 -0.714588   True  
614       0.034932 -0.712964   True  
1242      0.010202 -0.682996   True

In [24]:
# largest score increases
merged.sort_values('delta', ascending=False).head()

comment  \
586     =Jewish or German?=  So, since Hefner is a G...   
1029                               for my Jewish tastes   
767                                into a Jewish family   
1021              , negative reference to Jewish people   
181     Don't you think here you guys prove there is...   

                                         comment_fuzzed     score  \
586     =gay or German?=  So, since Hefner is a Germ...  0.000155   
1029                                  for my gay tastes  0.041840   
767                                   into a gay family  0.015356   
1021          , negative reference to homosexual people  0.005351   
181     Don't you think here you guys prove there is...  0.008182   

      score_fuzzed     delta  toxic  
586       0.726115  0.725960  False  
1029      0.757154  0.715314  False  
767       0.704413  0.689057  False  
1021      0.640352  0.635001  False  
181       0.532344  0.524162  False

### Equality of Opportunity

To evaluate equality of opportunity, you have to compare the false negative rates on a per-term basis.

In [25]:
def compute_negative_rates(df, score_col, label_col, threshold = 0.5):
        confusion_matrix = confusion_matrix_counts(df, score_col, label_col, threshold)
        tnr = confusion_matrix['tn']/(confusion_matrix['tn'] + confusion_matrix['fp'])
        fnr = confusion_matrix['fn']/(confusion_matrix['fn'] + confusion_matrix['tp'])
        return tnr, fnr

In [26]:
# per-term true negative rate

def per_term_negative_rates(df, term, text_col, label_col, score_col, threshold = 0.5):
    term_df = df[df[text_col].str.contains(r'\b{}\b'.format(term), case=False)]
    tnr, fnr = compute_negative_rates(term_df, score_col, label_col, threshold)
    return {'size': len(term_df),
            'term': term,
            'tnr': tnr,
            'fnr': fnr}

def terms_negative_rates(df, terms, text_col='Text', label_col='label', score_col='score'):
    return [per_term_negative_rates(df, term, text_col, label_col, score_col)
            for term in terms]

In [27]:
def print_terms_negative_rates():
    with open(MADLIBS_TERMS_PATH) as f:
        madlibs_terms = [term.strip() for term in f.readlines()]

    for rec in sorted(terms_negative_rates(madlibs, madlibs_terms), key=lambda d: d['tnr']):
        size, term, tnr, fnr = rec['size'], rec['term'], rec['tnr'], rec['fnr']
        print('{:18s} TNR: {:.4f}\tFNR: {:.4f}\texamples: {} ({:.1f}%)'.format(
            term, tnr, fnr, size, 100 * (size / len(madlibs)))) 

print_terms_negative_rates()

gay                TNR: 0.0199	FNR: 0.0000	examples: 1865 (2.1%)
homosexual         TNR: 0.4035	FNR: 0.0011	examples: 1814 (2.0%)
queer              TNR: 0.8025	FNR: 0.0128	examples: 1901 (2.1%)
lesbian            TNR: 0.9905	FNR: 0.1408	examples: 1912 (2.1%)
black              TNR: 0.9932	FNR: 0.3350	examples: 1868 (2.1%)
heterosexual       TNR: 0.9934	FNR: 0.1998	examples: 1858 (2.1%)
teenage            TNR: 0.9935	FNR: 0.2909	examples: 1877 (2.1%)
transgender        TNR: 0.9937	FNR: 0.2503	examples: 1912 (2.1%)
latino             TNR: 0.9939	FNR: 0.2996	examples: 1920 (2.1%)
sikh               TNR: 0.9946	FNR: 0.2978	examples: 1906 (2.1%)
deaf               TNR: 0.9946	FNR: 0.3279	examples: 1837 (2.1%)
bisexual           TNR: 0.9946	FNR: 0.3273	examples: 1872 (2.1%)
middle aged        TNR: 0.9946	FNR: 0.3921	examples: 1893 (2.1%)
muslim             TNR: 0.9947	FNR: 0.3313	examples: 1924 (2.2%)
elderly            TNR: 0.9947	FNR: 0.3830	examples: 1870 (2.1%)
protestant         TNR: 0